### PR TITLE
feat: implement --build_python_zip pex

### PR DESCRIPTION
**Pull request by thesayyn**
### Type of change

- New feature or functionality (#236)

### Test plan

- New test cases added

### CODE REVIEW

File: MODULE.bazel
Comment (thesayyn) [inline MODULE.bazel:11]: This is needed to get tar toolchain.

**Comment by thesayyn** [comment]
This is kind of working now, 

Missing things i can think of;

- inject env provided via `ctx.attr.env`
- set runfiles directory so the runfiles lookup works.
- bzlmod runfiles mapping is not working. 
- only workspace relative imports (`import wksp.xyz`) and third-party (`import 3p.xyz`) imports work, we need to find a way to inject more entries into sys.path without using venv.

File: py/private/py_pex.bzl
Comment (dizzy57) [inline py/private/py_pex.bzl:37]:
```suggestion
    args.add("/usr/bin/env python3", format="--python-shebang=%s")
```

nit: `--python-shebang`'s help line says to pass shebang without `#!` (it works in both cases, whether you pass it or not)

File: py/private/py_binary.bzl
Comment (mattem) [inline py/private/py_binary.bzl:9]: Can the convention to rename on import be followed for this? _build_pex.

File: py/tools/pex/BUILD.bazel
Comment (mattem) [inline py/tools/pex/BUILD.bazel:7]: Where does the user get this dependency from? 

rules_py currently doesn't have any Python dependencies, this will be the first, so there is no call to a pip install in the setup. This is working here as it's using the internal test time dependencies that rules_py uses.

Ideally we wouldn't have a Python dependency at all, is there another way to interact with the PEX library, can we vendor the dependency?

**Review by mattem — CHANGES_REQUESTED** [review]

File: py/tools/pex/BUILD.bazel
Comment (thesayyn) [reply py/tools/pex/BUILD.bazel:27]: You are right, my latest commit fixes this by introducing a slightly complicated bootstrapping logic to allow pex builder to be either a py_binary from rules_python with a pip dependency (for local dev only) or a py_binary from rules_python with no external dependencies but a  pex builder packaged into a pex zip to prevent a python dependency.

Comment (thesayyn) [reply py/tools/pex/BUILD.bazel:27]: > Ideally we wouldn't have a Python dependency at all, is there another way to interact with the PEX library, can we vendor the dependency?

vendoring is another approach we could take, it's open for a debate.

File: py/private/py_pex.bzl
Comment (thesayyn) [reply py/private/py_pex.bzl]: I will explain this in detail once it is close to the finish line. i am not sure if this is the way to install third-party deps yet. I'll wait for some feedback from some internal users first.

File: py/tools/pex/BUILD.bazel
Comment (mattem) [inline py/tools/pex/BUILD.bazel:62]: There's an existing mechanism to select between the release and dev version of tools by switching on the `IS_PRERELEASE` symbol in `tools/version.bzl`, can we hook into the same structure here, rather than adding something new?

File: py/tools/pex/BUILD.bazel
Comment (mattem) [inline py/tools/pex/BUILD.bazel:4]: Perhaps rename one symbol explicitly to `rules_python_py_binary`? It's hard to follow with just the _ rename.

File: py/tools/pex/BUILD.bazel
Comment (mattem) [reply py/tools/pex/BUILD.bazel:17]: Given that pex is a single wheel with no dependencies, I'm not clear why we need to bootstrap it via a zip? 
Could we instead fetch the wheel via an `htt_file`, have rules_py expand it with the existing tools, then depend on that? rules_py has a fetch dependencies function where it declares other rule dependencies, so we could add the `http_file` into there.

Comment (thesayyn) [reply py/tools/pex/BUILD.bazel:17]: Yeah, that sounds better, i'll do that once get back to this again.

File: py/tools/pex/BUILD.bazel
Comment (thesayyn) [reply py/tools/pex/BUILD.bazel:62]: with the http_file approach, we don't need this as the pex builder always will be the same target.

**Comment by dizzy57** [comment]
I was unable to make it work with bzlmod, here's a minimal repro: https://github.com/dizzy57/bazel_py_minimal

File: py/tools/pex/BUILD.bazel
Comment (thesayyn) [reply py/tools/pex/BUILD.bazel]: This is now fixed.

File: examples/py_pex_binary/BUILD.bazel
Comment (mattem) [inline examples/py_pex_binary/BUILD.bazel:18]: Can these be tidied? I think I've commented on empty space in this diff before, perhaps a pass over would be good.

File: py/private/py_pex_binary.bzl
Comment (mattem) [inline py/private/py_pex_binary.bzl:59]: Does this need to capture the closure? Seems like it only references the var in the lambda.

File: py/private/py_pex_binary.bzl
Comment (mattem) [inline py/private/py_pex_binary.bzl:87]: What is the right thing to do? Should we create an issue rather than an TODO in the code that will never be todone?

File: py/repositories.bzl
Comment (mattem) [inline py/repositories.bzl:73]: Do you think we should namespace this more? `rules_py_pex_231` for example?

File: py/repositories.bzl
Comment (mattem) [inline py/repositories.bzl:70]:
```suggestion
    # The url and digest information can be found at https://pypi.org/pypi/pex/json
```

File: py/private/py_pex_binary.bzl
Comment (thesayyn) [reply py/private/py_pex_binary.bzl:87]: i must address this before landing this PR.

File: py/private/py_pex_binary.bzl
Comment (thesayyn) [reply py/private/py_pex_binary.bzl:59]: nope, left over. will remove

**Comment by thesayyn** [comment]
Also, CPython=={major}.{minor}.* should be the default constraint.

File: py/private/py_pex_binary.bzl
Comment (thesayyn) [reply py/private/py_pex_binary.bzl]: done

**Comment by thesayyn** [comment]
@mattem I have removed some unrelated diff from my failed attempt before using PEX.  Now diff should be smaller and easier to understand.

File: py/private/py_pex_binary.bzl
Comment (dizzy57) [inline py/private/py_pex_binary.bzl:123]:
```suggestion
    "binary": attr.label(executable = True, mandatory = True, cfg = "target"),
```
Building a pex without a `binary` doesn't make much sense.

Comment (njlr) [reply py/private/py_pex_binary.bzl:123]: I think that a PEX can be used as an environment only - IIUIR the entry point(s) are option in the PEX tool https://docs.pex-tool.org/buildingpex.html#specifying-entry-points

File: py/private/py_pex_binary.bzl
Comment (dizzy57) [inline py/private/py_pex_binary.bzl:16]: > `rules_python~~python~`

Will this path change if we set `--incompatible_use_plus_in_repo_names`? It's going to be set to true by default in bazel 8.

context: https://github.com/bazelbuild/bazel/issues/23127

File: py/private/py_pex_binary.bzl
Comment (dizzy57) [reply py/private/py_pex_binary.bzl:123]: @njlr there is no `deps` attribute in this rule (and I don't think we need it), so setting `binary` is the only way to specify what goes into PEX. In any way, the rule currently fails if you don't specify `binary` attribute:
```
ERROR: [...]/cli_tool/BUILD.bazel:19:14: in py_pex_binary rule //cli_tool:empty_pex:
Traceback (most recent call last):
	File "[...]/external/aspect_rules_py~/py/private/py_pex_binary.bzl", line 61, column 22, in _py_python_pex_impl
		runfiles = binary[DefaultInfo].data_runfiles
Error: type 'NoneType' has no operator [](Provider)
```

File: py/private/py_pex_binary.bzl
Comment (thesayyn) [reply py/private/py_pex_binary.bzl:16]: Yes, this will stop working.

Comment (thesayyn) [reply py/private/py_pex_binary.bzl:16]: Sent a filter for plus version as well. LMK if that does it.

File: py/private/py_pex_binary.bzl
Comment (thesayyn) [reply py/private/py_pex_binary.bzl:123]: done

File: py/private/py_pex_binary.bzl
Comment (thesayyn) [reply py/private/py_pex_binary.bzl:76]: Please see the comment below the argument. 

https://github.com/aspect-build/rules_py/pull/324/files#diff-bdbfb557a8b870d00314cf49db0fad2611b37d9d20d6c055034aef05b30a6e65R77

File: py/private/py_pex_binary.bzl
Comment (thesayyn) [reply py/private/py_pex_binary.bzl:18]: It changes based on `--incompatible_use_plus_in_repo_names` flag. Left a comment above them.

File: py/private/py_pex_binary.bzl
Comment (thesayyn) [reply py/private/py_pex_binary.bzl:80]: > # this is needed to allow passing a lambda (with workspace_name) to map_each

Comment (mattem) [reply py/private/py_pex_binary.bzl:80]: But this isn't that call, the `map_each` here is `lambda e: "--inject-env=%s=%s" % (e[0], e[1]),` so how is the call below related?

Comment (dizzy57) [reply py/private/py_pex_binary.bzl:80]: Lambda for `inject_env`'s `map_each` does not contain any references to the `workspace_name`.

Comment (thesayyn) [reply py/private/py_pex_binary.bzl:80]: i could get rid of it by creating a global `def`, this feels more localized and i did not see any performance degradation.

File: py/private/py_pex_binary.bzl
Comment (dizzy57) [inline py/private/py_pex_binary.bzl:122]: Do we need to return `python_zip_file` output group here?

Looks like `py_pex_binary` is a rule on it's own  by now and does not need any output groups. I think it's fine to return just the `DefaultInfo` provider.

File: py/private/py_pex_binary.bzl
Comment (dizzy57) [reply py/private/py_pex_binary.bzl:80]: https://bazel.build/rules/lib/builtins/Args#parameters_1 says
> To avoid unintended retention of large analysis-phase data structures into the execution phase, the map_each function must be declared by a top-level def statement; [...]

so I think it's better to rewrite it as a global `def`

Comment (mattem) [reply py/private/py_pex_binary.bzl:80]: In that case, can the comment be updated, as it's unrelated to what is happening in his call.

File: py/private/py_pex_binary.bzl
Comment (thesayyn) [reply py/private/py_pex_binary.bzl:122]: you are right, it's not necessary.

File: py/private/py_pex_binary.bzl
Comment (thesayyn) [reply py/private/py_pex_binary.bzl:80]: fixed the comment.

File: py/private/py_pex_binary.bzl
Comment (thesayyn) [reply py/private/py_pex_binary.bzl]: removed.

**Review by mattem — APPROVED** [review]

**Pull request merged**
